aten/src/ATen/native/TensorFactories.cpp

```diff
@@ -1096,6 +1096,128 @@ Tensor clone(const Tensor& src, c10::optional<c10::MemoryFormat> optional_memory
   return self;
 }
 
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ covariations ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Tensor cov(const Tensor& self, 
+           const c10::optional<Tensor>& otherOpt,
+           c10::optional<bool> rowvarOpt, 
+           c10::optional<bool> biasOpt, 
+           c10::optional<int64_t> ddofOpt,
+           const c10::optional<Tensor>& fweightsOpt, 
+           const c10::optional<Tensor>& aweightsOpt) {
+  
+  const Tensor& other = c10::value_or_else(otherOpt, [] {return Tensor();});
+  const Tensor& fweights = c10::value_or_else(fweightsOpt, [] {return Tensor();});
+  const Tensor& aweights = c10::value_or_else(aweightsOpt, [] {return Tensor();});
+  auto bias = biasOpt.value_or(false);
+  auto ddof = ddofOpt.value_or(0);
+  auto rowvar = rowvarOpt.value_or(true);
+
+  if(self.size(0) == 0)
+    return at::empty_like(self);  
+  TORCH_CHECK(self.dim() <= 2,
+              "Input1 has more than 2 dimensions." );
+    
+  Tensor X, y;
+  if(self.dim() == 1)
+    X = self.unsqueeze(0);
+  else
+    X = self;
+
+  if(!rowvar && X.size(0) != 1)
+    X = X.t();
+
+  if(other.defined()) {
+    TORCH_CHECK(other.dim() <= 2,
+                "Input2 has more than 2 dimensions." );
+ 
+    if(other.dim() == 1)
+      y = other.unsqueeze(0);
+    else
+      y = other;
+
+    if(!rowvar && y.size(0) != 1)
+      y = y.t();
+
+    std::vector<Tensor> t_list;
+    t_list.push_back(X); t_list.push_back(y);
+    X = at::cat(t_list, 0);
+  }
+
+  if (ddofOpt == c10::nullopt) {
+    ddof = (bias == false) ? 1 : 0;
+  }
+  
+  Tensor w = at::empty(0,self.options());
+  
+  if(fweights.defined()) {
+    TORCH_CHECK((fweights.scalar_type() == at::kLong || fweights.scalar_type() == at::kInt),
+                "fweights must be integer.");
+    TORCH_CHECK(fweights.dim() <= 1,
+                "cannot handle multidimensional fweights.");
+    TORCH_CHECK(fweights.size(0) == X.size(1),
+                "incompatible numbers of samples and weights.");
+ 
+    for (auto i =0; i<fweights.size(0); i++){
+      if (fweights[i].item<int>() < 0)
+        TORCH_CHECK(false,
+                  "fweights cannot be negative.");
+    }
+    w = at::clone(fweights);
```

Comment:
there seemed to be some memory leak when I tried just w = fweights, hence had cloned it.
Do you suggest something like ?
```
w = at::empty_like(fweights).copy_(fweights); 
```

Reply:
No, there shouldn't be a memory leak, can you show a code where `w=fweights` causes it? Pytorch codebase is full of assignments like this.

Reply:
So this is a minimal example of when i use `w=fweights`:

test used:
```python

m = torch.randn((10,), dtype = float)
n = torch.randn((10,), dtype = float)
f = torch.randint(0,10,(10,))*2
a = torch.randint(0,10,(10,))**2
rowvar = True
ddof = None
bias = True

print(np.cov(m.numpy(), n.numpy(), fweights = f.numpy(),
 aweights = a.numpy(), ddof = ddof, bias=bias))

print('-'*10)
print("value of f before ",f,"\n")

print(torch.cov(m, n,fweights = f, aweights = a, ddof = ddof, bias = bias))

print("\n")
print("value of f after ",f)
print('-'*10)

print(np.cov(m.numpy(), n.numpy(), fweights = f.numpy(),
 aweights = a.numpy(), ddof = ddof, bias=bias))

```

output:

```
[[0.34098991 0.36849235]
 [0.36849235 1.44821068]]
----------
value of f before  tensor([10, 14, 16, 18, 16,  8, 12, 12,  6, 14]) 

tensor([[0.3410, 0.3685],
        [0.3685, 1.4482]], dtype=torch.float64)


value of f after  tensor([  0, 896,   0,  72, 256, 288, 768,  12,  24,   0])
----------
[[0.19665162 0.45899477]
 [0.45899477 1.48449604]]


```

Basically in some way, fweights is getting changed to `fweights*aweights` when the operation `w *=aweights` happens. This does not happen when I clone or copy fweights to w instead of `w=fweights`. Seems strange, am I doing something wrong, @ngimel?

Reply:
```
w = fweights;
w *= aweights;
```
is expected to change fweights - first you assign fweights to w, then inplace modify w. To avoid this, you can do
```
w=fweights;
w = w*aweights;
```
This avoids extra operations that `clone` does, and also does not change values in fweights.

Reply:
Ahh I see! Thank you very much for the insights 😁.